[All - Fix] Improved threading safety, notably on macOS

Kodikuu opened the pull request:
When destroying threading-related contexts, the NULL assignment is moved to be first.
When calling other threading-related functions, the context is NULL-checked first.

This should help prevent some rare and specific race conditions around the calling of threading functions during or after the destruction of their contexts.

Kodikuu committed:
MTY_Mutex safety

Kodikuu committed:
MTYCond safety

Kodikuu committed:
MTY_RWLock safety

Kodikuu committed:
MTY_Waitable safety

Kodikuu committed:
MTY_ThreadPool safety

Kodikuu committed:
pre-push testing

bmcnett commented on src/thread.c:
```diff
@@ -164,16 +176,19 @@ void MTY_WaitableDestroy(MTY_Waitable **waitable)
 		return;
 
 	MTY_Waitable *ctx = *waitable;
```
the optimizer is free to ruin the change you're trying to make here. to be guaranteed the results you seek, you should do an `exchange` here. i don't think matoya has `exchange` so a `CAS` will do. but, matoya has no `CAS` of a pointer, which is odd, because that's the first kind of CAS an atomics library usually gets.

Kodikuu replied:
This would need combining with https://github.com/parsec-cloud/libmatoya/pull/131

Kodikuu replied:
Integrated memory barriers

Kodikuu committed:
Mistake pointed out by Julien

Namaneo committed:
Clear OBJC_CTX after connection close

Kodikuu committed:
Add memory barrier

bmcnett commented on src/thread.c:
```diff
@@ -270,6 +291,9 @@ static void *thread_pool_func(void *opaque)
 
 uint32_t MTY_ThreadPoolDispatch(MTY_ThreadPool *ctx, MTY_AnonFunc func, void *opaque)
 {
+	if (!ctx)
```
should we either emit an MTY_Log warning specific to this new error (NULL context being passed) or emit the same MTY_Log warning we do when we can't find an available index?

bmcnett commented on src/windows/threadw.c:
```diff
@@ -116,25 +116,35 @@ void MTY_MutexDestroy(MTY_Mutex **mutex)
 bool MTY_MutexTryLock(MTY_Mutex *ctx)
 {
+	if (!ctx)
+		return false;;
```
two semicolons here ;;

bmcnett reviewed (approved):
eventually, we would want to emit a log warning for these extraordinary cases, where the code expects to do a thing through a pointer, and the pointer is unexpectedly NULL. handy if we had a WARNING macro that logged the source file name & line number, then it wouldn't take much creativity to come up with a unique log string

Kodikuu committed:
Extra semicolon

bmcnett committed:
for cross-platform code to use a memory barrier, there must be a function (not just a macro)

for cross-platform code to use a memory barrier, there must be a function (not just a macro)

Namaneo commented:
A follow-up PR from @bmcnett changes the memory barrier to be a function instead of a macro, depending on when this PR is intended to be merged you might want to update it to reflect the upcoming changes.

- https://github.com/parsec-cloud/libmatoya/pull/138

Kodikuu committed:
Switch to MTY_MemoryBarrier

Kodikuu commented:
Rebased onto new branch and adjusted to use new function. No further changes

Namaneo commented on src/thread.c:
```diff
@@ -55,17 +55,21 @@ void MTY_RWLockDestroy(MTY_RWLock **rwlock)
 
 	MTY_RWLock *ctx = *rwlock;
+	*rwlock = NULL;
+	MTY_MemoryBarrier();
```
Just as a wild thought, it would be neat to have an `MTY_SecureNull(...)` function, which would have a this definition:
```c
void *MTY_SecureNull(void **ptr)
{
	void *old_ptr = *ptr;
	*ptr = NULL;
	MTY_MemoryBarrier();
	return old_ptr;
}
```
And taking that block as an example, you'd simply do instead: `MTY_RWLock *ctx = MTY_SecureNull(rwlock);`

_Disclaimer: I don't know how memory barriers work internally, i.e. if they're scoped to the executing function or not. If they are, my suggestion just make no sense, please ignore._

Kodikuu replied:
What you're actually after here is an atomic pointer swap, which Bryan does have code for. But I just want to get this work in, because it covers the main issue. Atomically swapping the pointer is less of an improvement (but still one)

See https://github.com/parsec-cloud/libmatoya/commits/thread-safety-barrier/

Namaneo reviewed (approved):
Approved with a single suggestion, that might not even be related to this very PR, but I figured I'd still write it here since their are examples to illustrate what I'm saying.

bmcnett committed:
Revert "Switch to MTY_MemoryBarrier"

This reverts commit cf5426256850c2f22d9a9199f24787f45deab7cc.

bmcnett committed:
not yet, for this thing

not yet, for this thing

bmcnett committed:
remnants of the thing we don't want here

remnants of the thing we don't want here

bmcnett committed:
and, the spacing

and, the spacing
